File: raffia-core/src/main/java/com/codepoetics/raffia/writers/PassThroughWriter.java
package com.codepoetics.raffia.writers;

import com.codepoetics.raffia.api.BasketWriter;

import java.math.BigDecimal;

public abstract class PassThroughWriter<T extends BasketWriter<T>, S extends PassThroughWriter<T, S>> implements BasketWriter<S> {

  protected final T state;

  protected PassThroughWriter(T state) {
    this.state = state;
  }

  protected abstract S with(T state);

  @Override
  public S beginObject() {
    return with(state.beginObject());
  }

  @Override
  public S beginArray() {
    return with(state.beginArray());
  }

  @Override
  public S end() {
    return with(state.end());
  }

  @Override
  public S key(String key) {
    return with(state.key(key));
  }

  @Override
  public S add(String value) {
    return with(state.add(value));
  }

  @Override
  public S add(BigDecimal value) {
    return with(state.add(value));
  }

  @Override
  public S add(boolean value) {
    return with(state.add(value));
  }

  @Override
  public S addNull() {
    return with(state.addNull());
  }

}

File: raffia-core/src/main/java/com/codepoetics/raffia/writers/PassThroughWriter.kt
package com.codepoetics.raffia.writers

import java.math.BigDecimal

abstract class PassThroughWriter<T : BasketWriter<T>, S : PassThroughWriter<T, S>> protected constructor(protected val state: T) : BasketWriter<S> {

    protected abstract fun with(state: T): S

    override fun beginObject(): S = with(state.beginObject())

    override fun beginArray(): S = with(state.beginArray())

    override fun end(): S {
        return with(state.end())
    }

    override fun key(key: String): S = with(state.key(key))

    override fun add(value: String): S = with(state.add(value))

    override fun add(value: BigDecimal): S = with(state.add(value))

    override fun add(value: Boolean): S = with(state.add(value))

    override fun addNull(): S = with(state.addNull())

}
